pkg/dio/multi_test.go

```diff
@@ -12,365 +12,278 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Package io adds dynamic behaviour to the standard io package mutliX types
 package dio
 
 import (
 	"bytes"
 	"io"
-	"sync"
+	"strings"
 	"testing"
+)
+
+const (
+	count = 32
 
-	log "github.com/Sirupsen/logrus"
+	base    = "base functionality"
+	dynamic = "dynamic add/remove functionality"
 )
 
-func init() {
-	// enable verbose logging during tests
-	log.SetLevel(log.DebugLevel)
-	verbose = true
+func write(t *testing.T, mwriter DynamicMultiWriter, p []byte) {
+	n, err := mwriter.Write(p)
+	if err != nil {
+		t.Errorf("write: %v", err)
+	}
+	if n != len(p) {
+		t.Errorf("short write: %d != %d", n, len(p))
+	}
 }
 
-func TestMultiWrite(t *testing.T) {
-	pipeAR, pipeAW := io.Pipe()
-	pipeBR, pipeBW := io.Pipe()
+func read(t *testing.T, mreader DynamicMultiReader, limit int) []byte {
+	total := 0
 
-	mwriter := MultiWriter(pipeAW, pipeBW)
+	var buf = make([]byte, 32*1024)
+	for {
+		n, err := mreader.Read(buf)
```

Comment:
think this is supposed to be a slice:
`buf[total:limit]`

if you don't have the first you're overwriting data, if you don't have the latter you may read past limit in the reader and lose data against a subsequent call.